Попытка работы с интерфесом

rick1177 opened the pull request.

rick1177 committed:
Попытка работы с интерфесом

navferty commented on ConsoleApp1/Program.cs:
```diff
@@ -36,7 +36,7 @@ static void Main(string[] args)
+    public static void GetList<T>(List<T> m_list) where T: IPrintForClasses
     {
         if (m_list is null) return;
         if ((m_list is List<Factories> || m_list is List<Units> || m_list is List<Tanks>) /*&& m_list is not null*/)
```
эти условия на конкретные типы не нужны. благодаря ограничению `where T: IPrintForClasses` этот метод знает, что у элементов в листе есть нужный метод.
подробнее можно почитать про Принцип подстановки Барбары Лисков (буква L в аббревиатуре [SOLID](https://ru.wikipedia.org/wiki/SOLID_(объектно-ориентированное_программирование)))

navferty commented on ConsoleApp1/Program.cs:
```diff
@@ -50,7 +50,7 @@ static void Main(string[] args)
-class Factories //Завод
+class Factories : IPrintForClasses//Завод
 {
     public string name = "Без названия";    // наименование
```
Здесь и ниже рекомендую использовать автосвойства вместо публичных полей. Вот [тут](https://ru.stackoverflow.com/a/197075) неплохо расписано почему

navferty replied:
Также публичные свойства обычно именуют в PascalCase - то есть с большой буквы

rick1177 committed:
Корректный вариант работы с интерфейсом

rick1177 committed:
Внесены исправления:
- вместо полей используются свойста;
- убраны лишние проверки в методе GetList

navferty commented on ConsoleApp1/Program.cs:
```diff
@@ -36,56 +36,90 @@ static void Main(string[] args)
-    public static void GetList<T>(T m_list) where T: IList
+    public static void GetList<T>(List<T> list) where T: IPrintForClasses
     {
-        if (m_list is null) return;
```
а вот проверка на null как раз будет нелишней, обезопасит от NullReferenceException

navferty commented on ConsoleApp1/Program.cs:
```diff
@@ -36,56 +36,90 @@ static void Main(string[] args)
+    private string name = "Без названия";    // наименование
+    public string Name
+    {
+        get
```
Нет необходимости явно прописывать backing field, в таких случаях можно обходиться автосвойствами, здесь и ниже.

Примерно так:
```c#
public string Name { get; set; } = "Без названия";
```

rick1177 committed:
Внесены изменения:
- упрощён синтаксис для свойств Name, обходимся автосвойствами;
- возвращена проверка на null в функции GetList
